internal/cli/shell.go

```diff
@@ -0,0 +1,202 @@
+package cli
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"syscall"
+
+	"github.com/spf13/cobra"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/tilt-dev/tilt/internal/container"
+	"github.com/tilt-dev/tilt/internal/controllers/core/kubernetesdiscovery"
+	"github.com/tilt-dev/tilt/internal/k8s"
+	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
+	"github.com/tilt-dev/tilt/pkg/logger"
+	"github.com/tilt-dev/tilt/pkg/model"
+)
+
+type shellCmd struct {
+	streams   genericclioptions.IOStreams
+	container string
+	execer    ShellExecer
+}
+
+func newShellCmd(streams genericclioptions.IOStreams) *shellCmd {
+	return &shellCmd{
+		streams: streams,
+		execer:  realShellExecer{},
+	}
+}
+
+func (c *shellCmd) name() model.TiltSubcommand { return "shell" }
+
+func (c *shellCmd) register() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:                   "shell [<resource-name>]",
+		DisableFlagsInUseLine: true,
+		Short:                 "Opens a shell into a container running in Tilt",
+		Long: `Opens a shell into a container running in Tilt.
+
+Given a resource name, finds the Kubernetes Pod in that resource,
+and opens an interactive shell.
+
+By default, in order, we'll try:
+- kubectl shell
+- dld shell
+- docker exec -it <container> bash
+- docker exec -it <container> sh`,
```

Comment:
should there be a fallback to `kubectl debug` also?

Reply:
i think `kubectl debug` is for a different use-case? i.e., it's more for "i'm debugging the infra and want to ssh to the Node where this pod lives and check the firewall rules" rather than tilt's "i'm debugging the container and want to poke around on its filesystem"

Reply:
It's interesting you thought that. You can attach to a node, but you can do the same with a pod. It's currently a bit short of functionality of enabling what 'kubectl shell' does on the client side, there is an open PR to add a privileged profile.